File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/PubliclyWritableDirectories.cs
using System;
using System.IO;

namespace Tests.Diagnostics
{
    public class Program
    {
        public void Compliant()
        {
            var randomPath = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName()); // Compliant
            var tempFileName = Path.GetTempFileName(); // Compliant

            using (var tmpDir = new StreamReader(@"Path is /tmp/f")) { };  // Compliant
            using (var tmpDir = new StreamReader(@"C:\\Windows\\Tempete")) { };  // Compliant;
            var url = "http://example.domain/tmp/f"; // Compliant
        }

        private string CompliantWithArg(string dir = "/other/tmp") // Compliant
        {
            return dir;
        }

        private class InnerClass
        {
            public string StringProp { get; set; }
        }

        public void NonCompliant(string partOfPath)
        {
            // Environment
            var tmp = Path.GetTempPath();   // Noncompliant
            tmp = Path.GetTempPath();       // Noncompliant
            InnerClass inner = new InnerClass() { StringProp = Path.GetTempPath() }; // Noncompliant
            tmp = Environment.GetEnvironmentVariable("TMPDIR");                      // Noncompliant {{Make sure publicly writable directories are used safely here.}}
//                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
            tmp = Environment.GetEnvironmentVariable("TMP");            // Noncompliant
            tmp = Environment.GetEnvironmentVariable("TEMP");           // Noncompliant

            string temp = "TEMP";
            tmp = Environment.GetEnvironmentVariable(temp); // FN

            tmp = Environment.GetEnvironmentVariable(42); // Error [CS1503]
            tmp = Environment.GetEnvironmentVariable();   // Error [CS1501]

            tmp = Environment.ExpandEnvironmentVariables("%TMPDIR%");   // Noncompliant
            tmp = Environment.ExpandEnvironmentVariables("%TMP%");      // Noncompliant
            tmp = Environment.ExpandEnvironmentVariables("%TEMP%");     // Noncompliant
            tmp = "%USERPROFILE%\\AppData\\Local\\Temp\\f";             // Noncompliant
            tmp = "%TEMP%\\f";                                          // Noncompliant
            tmp = "%TMP%\\f";                                           // Noncompliant
            tmp = "%TMPDIR%\\f";                                        // Noncompliant
//                ^^^^^^^^^^^^^

            // Common
            using (var tmpDir = new StreamReader("/tmp/f")) { };        // Noncompliant
            using (var tmpDir = new StreamReader("/tmp")) { };          // Noncompliant
            using (var tmpDir = new StreamReader("/var/tmp/f")) { };    // Noncompliant
            using (var tmpDir = new StreamReader("/usr/tmp/f")) { };    // Noncompliant
            using (var tmpDir = new StreamReader("/dev/shm/f")) { };    // Noncompliant

            // Linux
            using (var tmpDir = new StreamReader("/dev/mqueue/f")) { };     // Noncompliant
            using (var tmpDir = new StreamReader("/run/lock/f")) { };       // Noncompliant
            using (var tmpDir = new StreamReader("/var/run/lock/f")) { };   // Noncompliant

            // MacOS
            using (var tmpDir = new StreamReader("/Library/Caches/f")) { };     // Noncompliant
            using (var tmpDir = new StreamReader("/Users/Shared/f")) { };       // Noncompliant
            using (var tmpDir = new StreamReader("/private/tmp/f")) { };        // Noncompliant
            using (var tmpDir = new StreamReader("/private/var/tmp/f")) { };    // Noncompliant

            // Windows
            using (var tmpDir = new StreamReader("C:\\Windows\\Temp\\f")) { };  // Noncompliant
            using (var tmpDir = new StreamReader("C:\\Temp\\f")) { };           // Noncompliant
            using (var tmpDir = new StreamReader("C:\\TEMP\\f")) { };           // Noncompliant
            using (var tmpDir = new StreamReader("C:\\TMP\\f")) { };            // Noncompliant

            // Variates
            using (var tmpDir = new StreamReader(@"/tmp/f")) { };                   // Noncompliant
            using (var tmpDir = new StreamReader("D:\\Windows\\Temp\\f")) { };      // Noncompliant
            using (var tmpDir = new StreamReader("\\\\Server_Name\\Temp\\f")) { };  // Noncompliant
            using (var tmpDir = new StreamReader("\\Windows\\Temp\\f")) { };        // Noncompliant
            using (var tmpDir = new StreamReader(@"C:\Windows\Temp\f")) { };        // Noncompliant
            using (var tmpDir = new StreamReader(@"D:\Windows\Temp\f")) { };        // Noncompliant
            using (var tmpDir = new StreamReader(@"\\Windows\Temp\f")) { };         // Noncompliant
            using (var tmpDir = new StreamReader(@"\Windows\Temp\f")) { };          // Noncompliant
            tmp = "/tmp/f";                                                         // Noncompliant
            tmp = "/tEmP/f";                                                        // Noncompliant

            // Interpolated
            tmp = $"/tmp/{partOfPath}";                                             // Noncompliant
//                ^^^^^^^^^^^^^^^^^^^^
        }

        private string NonCompliantWithArg(string dir = "/tmp") // Noncompliant
        {
            return dir;
        }

        public string TempPath => Path.GetTempPath();           // Noncompliant

        public string GetTempPath()
        {
            return Path.GetTempPath();                          // Noncompliant
        }
    }
}

File: analyzers/tests/SonarAnalyzer.Test/TestCases/Hotspots/PubliclyWritableDirectories.vb
Imports System
Imports System.IO

Namespace Tests.TestCases

    Public Class Program

        Public Sub Compliant()
            Dim RandomPath As String = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName()) ' Compliant
            Dim TempFileName As String = Path.GEtTempFileName() ' Compliant
            Using TmpDir = New StreamReader("Path is /tmp/f") ' Compliant
            End Using
            Using TmpDir = New StreamReader("C:\\Windows\\Tempete") ' Compliant;
            End Using
            Dim url As String = "http://example.domain/tmp/f" ' Compliant
        End Sub

        Public Function CompliantWithArg(Optional Dir As String = "/other/tmp") As String ' Compliant
            Return Dir
        End Function

        Public Sub NonCompliant(PartOfPath As String)
            ' Environment
            Dim Tmp As String = Path.GetTempPath()                      ' Noncompliant
            Tmp = Path.GetTempPath()                                    ' Noncompliant
            Tmp = Environment.GetEnvironmentVariable("TMPDIR")          ' Noncompliant {{Make sure publicly writable directories are used safely here.}}
            '     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
            Tmp = Environment.GetEnvironmentVariable("TMP")             ' Noncompliant
            Tmp = Environment.GetEnvironmentVariable("TEMP")            ' Noncompliant
            Tmp = Environment.GetEnvironmentVariable(42)
            Tmp = Environment.GetEnvironmentVariable() ' Error [BC30516]
            Tmp = Environment.ExpandEnvironmentVariables("%TMPDIR%")    ' Noncompliant
            Tmp = Environment.ExpandEnvironmentVariables("%TMP%")       ' Noncompliant
            Tmp = Environment.ExpandEnvironmentVariables("%TEMP%")      ' Noncompliant
            Tmp = "%USERPROFILE%\AppData\Local\Temp\f"                  ' Noncompliant
            Tmp = "%TEMP%\f"                                            ' Noncompliant
            Tmp = "%TMP%\f"                                             ' Noncompliant
            Tmp = "%TMPDIR%\f"                                          ' Noncompliant
            '     ^^^^^^^^^^^^

            ' Common
            Using TmpDir = New StreamReader("/tmp/f")           ' Noncompliant
            End Using
            Using TmpDir = New StreamReader("/tmp")             ' Noncompliant
            End Using
            Using TmpDir = New StreamReader("/var/tmp/f")       ' Noncompliant
            End Using
            Using TmpDir = New StreamReader("/usr/tmp/f")       ' Noncompliant
            End Using
            Using TmpDir = New StreamReader("/dev/shm/f")       ' Noncompliant
            End Using

            ' Linux
            Using TmpDir = New StreamReader("/dev/mqueue/f")    ' Noncompliant
            End Using
            Using TmpDir = New StreamReader("/run/lock/f")      ' Noncompliant
            End Using
            Using TmpDir = New StreamReader("/var/run/lock/f")  ' Noncompliant
            End Using

            ' MacOS
            Using TmpDir = New StreamReader("/Library/Caches/f")    ' Noncompliant
            End Using
            Using TmpDir = New StreamReader("/Users/Shared/f")      ' Noncompliant
            End Using
            Using TmpDir = New StreamReader("/private/tmp/f")       ' Noncompliant
            End Using
            Using TmpDir = New StreamReader("/private/var/tmp/f")   ' Noncompliant
            End Using

            ' Windows
            Using TmpDir = New StreamReader("C:\Windows\Temp\f") ' Noncompliant
            End Using
            Using TmpDir = New StreamReader("C:\Temp\f")         ' Noncompliant
            End Using
            Using TmpDir = New StreamReader("C:\TEMP\f")         ' Noncompliant
            End Using
            Using TmpDir = New StreamReader("C:\TMP\f")          ' Noncompliant
            End Using

            ' Variates
            Using TmpDir = New StreamReader("/tmp/f")               ' Noncompliant
            End Using
            Using TmpDir = New StreamReader("\\Server_Name\Temp\f") ' Noncompliant
            End Using
            Using TmpDir = New StreamReader("C:\Windows\Temp\f")    ' Noncompliant
            End Using
            Using TmpDir = New StreamReader("D:\Windows\Temp\f")    ' Noncompliant
            End Using
            Using TmpDir = New StreamReader("\Windows\Temp\f")      ' Noncompliant
            End Using
            Tmp = "/tmp/f"  ' Noncompliant
            Tmp = "/TeMp/f" ' Noncompliant

            ' Interpolated
            Tmp = $"/tmp/{PartOfPath}" ' Noncompliant
            '     ^^^^^^^^^^^^^^^^^^^^

        End Sub

        Public Function NonCompliantWithArg(Optional Dir As String = "/tmp") As String ' Noncompliant
            Return Dir
        End Function

        ReadOnly Property GetTempPath() As String
            Get
                Return Path.GetTempPath() ' Noncompliant
            End Get
        End Property

        Public Sub Examples()
            Const t As String = "T"
            Const e As String = "E"
            Const m As String = "M"
            Const p As String = "P"
            Const part1 As String = "/tEmP" ' Noncompliant
            Const part2 As String = "/f"
            Dim noncompliant2 As String = $"{part1}{part2}" ' Noncompliant
            Dim tmp = Environment.GetEnvironmentVariable($"{t}{e}{m}{p}") ' Noncompliant
            tmp = Environment.GetEnvironmentVariable($"{t}{e}{m}{p}{5}")
        End Sub

    End Class

End Namespace
